### PR TITLE
Add support to build macOS & Windows on Azure DevOps

### PR DESCRIPTION
**Pull request by junyzheng**
this patch was created in last December to try to replace #1306. There is no update since then. Probably need a little bit work to make this work.

### CODE REVIEW

File: .azure-devops/build/steps/macOS/pre.yml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+  # create all required directories before downloading files & installing jdks
+  - bash: |
+      mkdir -p ${{ parameters.jdkBootDir }}
+      mkdir -p ${{ parameters.javaHomeDir }}
+      mkdir -p ${{ parameters.dependenciesDir }}
+    displayName: "[macOS Pre] create required directories"
+
+  # download JDKs from AdoptOpenJDK
+  - bash: |
```
Comment (karianna) [inline .azure-devops/build/steps/macOS/pre.yml:48]: What about for Java 8?  Need to get a Java 7 boot.  Adopt doesn't have these, so I recommend Zulu

File: .azure-devops/build/steps/macOS/pre.yml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      then
+          BOOTJDK_VERSION=10
+      elif [ "$(JAVA_TO_BUILD)" == "jdk13u" ]
+      then
+          BOOTJDK_VERSION=12
+      elif [ "$(JAVA_TO_BUILD)" == "jdk" ]
+      then
+          BOOTJDK_VERSION=13
+      fi
+      wget -O "${{ parameters.dependenciesDir }}/openjdk8.tar.gz" "https://api.adoptopenjdk.net/v2/binary/releases/openjdk8?os=mac&release=latest&arch=x64&heap_size=normal&type=jdk&openjdk_impl=hotspot"
```
Comment (karianna) [inline .azure-devops/build/steps/macOS/pre.yml:60]: Please use v3 of the API

File: .azure-devops/build/steps/shared/pre.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    displayName: "[Pre] set VARIANT environment variables"
+
+  # check for required environment variables
+  # pipeline will be terminated if environment variables are not defined
+  # reference: https://docs.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands
+  - bash: |
+      ERR_MESSAGE_MISSING_ENV=''
+
+      if [ -z $(JAVA_TO_BUILD) ]; then
+        ERR_MESSAGE_MISSING_ENV="Pipeline must set JAVA_TO_BUILD to jdk8, jdk11, jdk, etc. $ERR_MESSAGE_MISSING_ENV"
```
Comment (karianna) [inline .azure-devops/build/steps/shared/pre.yml:19]: should those messages read jdk11 or jdk11u etc?

**Comment by junyzheng** [comment]
@karianna this patch was created in December/January. I was talking to @llxia and they are working on integration with AzDO (please correct me if I am wrong) and need to fix #1306. So here is just a something they can take a look (I think this script also need some fix), maybe use this instead of #1306. I will change this PR to draft.

**Comment by karianna** [comment]
> @karianna this patch was created in December/January. I was talking to @llxia and they are working on integration with AzDO (please correct me if I am wrong) and need to fix #1306. So here is just a something they can take a look (I think this script also need some fix), maybe use this instead of #1306. I will change this PR to draft.

OK, makes sense, can you check with @llxia and then close #1306 as a duplicate?

**Comment by llxia** [comment]
Sorry, I missed the above message. 
Thanks @junyzheng I can run it in my AzDO
https://github.com/AdoptOpenJDK/openjdk-build/pull/1306 contains tests portion. Will it be added in another PR?

**Comment by llxia** [comment]
@junyzheng if you can get the build part delivered, I can work on the test part. As we have [smart parallelization](https://github.com/AdoptOpenJDK/openjdk-tests/issues/1563) enabled in tests, we can use this feature to leverage AzDO machines. Thanks

**Comment by junyzheng** [comment]
@llxia sorry just saw this. would yo mind to tell me what the JDK I should support? I guess tip, 11, and 8?

**Comment by smlambert** [comment]
Lan is away for remainder of today, so I will answer yes, tip, 8 and 11 would be great!  thanks @junyzheng !

**Comment by junyzheng** [comment]
OK, I will work on this next week!

**Comment by junyzheng** [comment]
@karianna @smlambert @llxia ready to be reviewed again, sample build: https://dev.azure.com/zjunyuan/junyuanz/_build the failures are caused by I don't have AzDO billing set up for my personal account, so it can't upload to Artifacts

**Comment by llxia** [comment]
Thanks @junyzheng

**Comment by karianna** [comment]
> Thanks @junyzheng ! @karianna did your comments get fully addressed? I will leave it to your final approval for merge.
> 
> Note: This PR opens the way for @llxia to add the smart parallelization testing. :)

John and I are humming and hahing about this. The PR is good but it opens up a philosophical point.  Many of the variables and versions references in here are 'duplicates' of our main build scripts, which means.... refactor time?  Not sure yet, I'll chat with John further this week

**Comment by junyzheng** [comment]
@karianna I am not sure how Adopt is planing to use this script. If they can set environment variables through by using Jenkins then I can simplify those YAML file.

File: .azure-devops/README.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+## Requirements
+
+1. Azure DevOps Account
+
+If you don't have an Azure DevOps organization, you can start from
+[here][azdo_main]
+
+2. Required Pipeline Variables:
+
+    1. `JAVA_TO_BUILD`: jdk8u | jdk11u | jdk
```
Comment (karianna) [inline .azure-devops/README.md:23]: We should add `jdk14u` to this?

File: .azure-devops/build/build.yml [hunk 5]
```diff
@@ -0,0 +1,77 @@
+# Build Stages Template
+#
+# supported platform & versions:
+#    macOS: jdk11u, jdk13u, and jdk
```
Comment (karianna) [inline .azure-devops/build/build.yml:4]: I think we should remove `jdk13u` and add `jdk14u` for macOS

File: .azure-devops/build/build.yml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+#    windows: jdk8u, jdk11u, jdk13u, and jdk
```
Comment (karianna) [inline .azure-devops/build/build.yml:5]: Do we add `jdk14u` to windows?

File: .azure-devops/build/build.yml [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+#
+# required pipeline level variable(s):
+#    JAVA_TO_BUILD: jdk8u|jdk11u|jdk13u|jdk
```
Comment (karianna) [inline .azure-devops/build/build.yml:8]: I think we should remove `jdk13u` and add `jdk14u`

File: .azure-devops/build/build.yml [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+  - stage: build_windows_x64_hotspot_binaries
+    variables:
+      skipComponentGovernanceDetection: true
+      TARGET_OS: "windows"
+      ARCHITECTURE: "x64"
+    displayName: "Build Windows x64 HotSpot Binaries"
+    dependsOn: []
+    jobs:
+      - job: build_macOS
```
Comment (karianna) [inline .azure-devops/build/build.yml:49]: So this requires the build_macOS job to complete first?

File: .azure-devops/build/steps/macOS/before.yml [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  # create all required directories before downloading files & installing jdks
+  - bash: |
+      mkdir -p ${{ parameters.jdkBootDir }}
+      mkdir -p ${{ parameters.javaHomeDir }}
+      mkdir -p ${{ parameters.dependenciesDir }}
+    displayName: "[macOS Pre] create required directories"
+
+  # download JDKs from AdoptOpenJDK
+  - bash: |
+      BOOTJDK_VERSION=
```
Comment (karianna) [inline .azure-devops/build/steps/macOS/before.yml:49]: Do we need to cover the jdk14u cases here as well?

File: .azure-devops/build/steps/shared/before.yml [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+      git config --global core.autocrlf input
+    displayName: "[Pre] configure crlf option"
+
+  # self checkout openjdk-build repo
+  - checkout: self
+
+  # set the git authorization header for subsequent git pull/clone operations...
+  # <This step is unnecessary on your local workstation, can be approximated by using SSH URIs>
+  - bash: |
+      git config --global --add http.https://dev.azure.com.extraheader "AUTHORIZATION: bearer $(System.AccessToken)"
```
Comment (karianna) [inline .azure-devops/build/steps/shared/before.yml:53]: Extract this if it's internal to MSFT

File: .azure-devops/build/steps/windows/before.yml [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    displayName: "[Windows Pre] add cygwin to PATH"
+
+  # unzip downloaded JDK zip files to directories, bsdtar is installed by cygwin
+  - bash: |
+      bsdtar --strip-components=1 -xvf $(cygpath -u '${{ parameters.dependenciesDir }}\jdk8.zip') -C '${{ parameters.javaHomeDir }}'
+      bsdtar --strip-components=1 -xvf $(cygpath -u '${{ parameters.dependenciesDir }}\jdk.zip') -C '${{ parameters.jdkBootDir }}'
+    displayName: "[Windows Pre] extracting downloaded JDKs to directories"
+
+  # only install visual studio 2013 when build jdk8 and jdk11
+  - powershell: |
```
Comment (karianna) [inline .azure-devops/build/steps/windows/before.yml:64]: In the short term future, we'd like to move to vs2017 - but we can tackle that separately

**Comment by karianna** [comment]
@junyzheng - Can you address the comments and push the latest?

File: .azure-devops/build/steps/windows/before.yml (hunk 11 above)
Comment (junyzheng) [reply .azure-devops/build/steps/windows/before.yml:52]: that's great the machine comes with vs2017

**Comment by junyzheng** [comment]
@karianna ready to be reviewed again: https://dev.azure.com/zjunyuan/junyuanz/_build

**Pull request merged**